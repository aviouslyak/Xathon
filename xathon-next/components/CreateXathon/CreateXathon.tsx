import React, { useState } from "react";
import { useForm } from "react-hook-form";
import Input from "../Input/Input";
import SubmitButton from "../SubmitButton/SubmitButton";
import Textarea from "../TextArea/Textarea";
import {
  XathonFactoryReader,
  XathonFactoryWriter,
} from "../../services/contracts/XathonFactory";
import { AiOutlinePlusCircle } from "react-icons/ai";
import { CgSpinner } from "react-icons/cg";
import ErrorText from "../ErrorText/ErrorText";
import Router from "next/router";

interface FormValues {
  address: string;
  name: string;
  description: string;
  unit: string;
  minimumDeposit: number;
}

const CreateXathon: React.FC = () => {
  const [isSubmitting, setIsSubmitting] = useState(false);
  const [error, setError] = useState("");

  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<FormValues>({
    defaultValues: {
      address: "",
      name: "",
      unit: "",
      description: "",
      minimumDeposit: 0,
    },
  });

  return (
    <form
      className="w-1/2 p-4 flex flex-col justify-items-start items-start gap-1"
      onSubmit={handleSubmit(async (data) => {
        setIsSubmitting(true);
        const xathonFactoryWriter = new XathonFactoryWriter();
        const xathonFactoryReader = new XathonFactoryReader();
        await xathonFactoryWriter.setAddresses();
        try {
          await xathonFactoryWriter.deployXathon(data);
          const address = await xathonFactoryReader.getContractAddress(
            data.name
          );
          Router.push(`/${address}`);
        } catch (err) {
          console.error(err);
          setError(
            "Some error has occurred. Check that you confirmed the transaction in MetaMask, and that the contract name is unique"
          );
        }
        setIsSubmitting(false);
      })}
      data-testid="form"
      id="create-xathon"
    >
      <h3 className="font-semibold text-xl self-center">
        Create a new <span className="accent-text">Xathon</span> contract
      </h3>
      <ErrorText className="text-md">{error}</ErrorText>
      <Input
        type="text"
        {...register("address", {
          required: "*Address is required",
          pattern: {
            value: /0x[a-fA-F0-9]{40}/,
            message: "*Must by an ethereum address",
          },
        })}
        errorText={errors.address && errors.address.message}
        labelText="Donation recipient address"
        placeholder="Enter address"
        width="w-2/3"
      />
      <Input
        type="text"
        {...register("name", {
          required: "*Name is required",
          maxLength: {
            value: 50,
            message: "*Must be less than 50 characters in length",
          },
        })}
        errorText={errors.name && errors.name.message}
        labelText="Xathon name"
        placeholder="Enter name"
        width="w-2/3"
      />
      <Textarea
        placeholder="Enter description"
        className="block w-2/3 h-20"
        {...register("description")}
      />
      <Input
        type="text"
        {...register("unit", {
          maxLength: {
            value: 20,
            message: "*Must be less than 20 characters in length",
          },
        })}
        labelText="Unit for xathon"
        errorText={errors.unit && errors.unit.message}
        placeholder="miles"
        width="w-30"
      />
      <Input
        type="text"
        {...register("minimumDeposit", { valueAsNumber: true })}
        errorText={errors.minimumDeposit && errors.minimumDeposit.message}
        labelText="Minimum deposit"
        placeholder="0"
        width="w-20"
      />
      <SubmitButton form="create-xathon" className="self-center">
        {isSubmitting ? (
          <span>
            <CgSpinner className="h-5 w-5 mr-1 inline animate-spin" />
            Submitting
          </span>
        ) : (
          <span>
            Create
            <AiOutlinePlusCircle className="transition-transform transform group-hover:scale-125 ml-2 inline-block" />
          </span>
        )}
      </SubmitButton>
      <p className="text-sm self-center mt-1">
        Your request could take a while to complete
      </p>
    </form>
  );
};

export default CreateXathon;
