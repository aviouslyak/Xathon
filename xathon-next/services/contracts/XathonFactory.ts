import { web3ReadOnly } from "../web3";
import { XathonFactory as XathonFactoryType } from "../../types/web3-v1-contracts/XathonFactory";
const contractJSON = require("../../ethereum/contracts/XathonFactory.json");

const ADDRESSES = {
  rinkeby: "0x7d6a89f57cF666e613baA13a1cce354237b070b0",
};

export default class XathonFactory {
  static ReadOnlyXF: XathonFactoryType = new web3ReadOnly.eth.Contract(
    contractJSON.abi,
    ADDRESSES.rinkeby
  ) as any as XathonFactoryType;

  private WriteOnlyXF: XathonFactoryType;
  constructor() {
    this.WriteOnlyXF = null as any as XathonFactoryType;
  }
  static async getContractNames() {
    return await this.ReadOnlyXF.methods.getDeployedXathons().call();
  }

  static async getContractAddress(name: string) {
    return await this.ReadOnlyXF.methods.getAddress(name).call();
  }
}
