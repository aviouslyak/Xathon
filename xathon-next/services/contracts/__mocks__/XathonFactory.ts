import { TransactionReceipt } from "web3-core";
const ADDRESS = "0x0000000000000000000000000000000000000000";

export default class XathonFactory {
  static async getContractNames() {
    return await new Promise((resolve) => resolve(["foo", "boo", "zoo"]));
  }

  static async getContractAddress(name: string) {
    return await new Promise((resolve) => resolve(ADDRESS));
  }

  async setAddresses() {
    return;
  }

  async deployXathon() {
    return await new Promise((resolve) =>
      resolve("foo" as any as TransactionReceipt)
    );
  }
}
