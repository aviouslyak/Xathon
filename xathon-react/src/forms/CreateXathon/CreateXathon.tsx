import React from "react";
import { useForm } from "react-hook-form";
import Input from "../../components/Input/Input";
import SubmitButton from "../../components/SubmitButton/SubmitButton";
import Textarea from "../../components/TextArea/Textarea";
import WalkathonFactory from "../../services/contracts/xathonFactory";

interface FormValues {
  address: string;
  name: string;
  description: string;
  unit: string;
  minimumDeposit: number;
}

const CreateWalkathonFactory: React.FC = () => {
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<FormValues>({
    defaultValues: {
      address: "",
      name: "",
      description: "",
      minimumDeposit: 0,
    },
  });
  return (
    <form
      className="responsive-component-width p-4 flex flex-col justify-items-start items-start"
      onSubmit={handleSubmit(async (data) => {
        console.log("SUBMITTING");
        const receipt = await WalkathonFactory.deployXathon(data);
        console.log(`DONE SUBMITTING: receipt = \n${receipt}`);
      })}
    >
      <h3 className="font-semibold text-xl self-center">
        Create new <span className="accent-text">Xathon</span> contract
      </h3>
      <Input
        type="text"
        {...register("address", {
          required: "*Address is required",
          pattern: {
            value: /0x[a-fA-F0-9]{40}/,
            message: "*Must by an ethereum address",
          },
        })}
        errorText={errors.address && errors.address.message}
        labelText="Enter donation recipient address"
        placeholder="Enter address"
        width="w-2/3"
      />
      <Input
        type="text"
        {...register("name", {
          required: "*Name is required",
          maxLength: {
            value: 50,
            message: "*Must be less than 50 characters in length",
          },
        })}
        errorText={errors.name && errors.name.message}
        labelText="Enter xathon name"
        placeholder="Enter name"
        width="w-2/3"
      />
      <Textarea
        placeholder="Enter description"
        className="block w-2/3"
        {...register("description")}
      />
      <Input
        type="text"
        {...register("unit", {
          maxLength: { value: 20, message: "*Must be less than 20 characters" },
        })}
        labelText="Enter unit for xathon"
        placeholder="miles"
        width="w-30"
      />
      <Input
        type="text"
        {...register("minimumDeposit", { valueAsNumber: true })}
        errorText={errors.minimumDeposit && errors.minimumDeposit.message}
        labelText="Enter minimum deposit"
        placeholder="0"
        width="w-20"
      />
      <SubmitButton value="Create" className="self-center" />
    </form>
  );
};

export default CreateWalkathonFactory;
