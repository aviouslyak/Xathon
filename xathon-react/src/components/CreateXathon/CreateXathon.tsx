import React, { useState } from "react";
import { useForm } from "react-hook-form";
import Input from "../Input/Input";
import SubmitButton from "../SubmitButton/SubmitButton";
import Textarea from "../TextArea/Textarea";
import XathonFactory from "../../services/contracts/xathonFactory";
import { CgSpinner } from "react-icons/cg";
import ErrorText from "../ErrorText/ErrorText";

interface FormValues {
  address: string;
  name: string;
  description: string;
  unit: string;
  minimumDeposit: number;
}

interface CreateXathonProps {
  setAddress: React.Dispatch<React.SetStateAction<string>>;
}
const CreateXathon: React.FC<CreateXathonProps> = ({ setAddress }) => {
  const [isSubmitting, setIsSubmitting] = useState(false);
  const [error, setError] = useState("");

  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<FormValues>({
    defaultValues: {
      address: "",
      name: "",
      unit: "",
      description: "",
      minimumDeposit: 0,
    },
  });

  return (
    <form
      className="responsive-component-width p-4 flex flex-col justify-items-start items-start"
      onSubmit={handleSubmit(async (data) => {
        setIsSubmitting(true);
        try {
          await XathonFactory.deployXathon(data);
          const address = await XathonFactory.getAddress(data.name);
          setAddress(address);
        } catch (err) {
          setError(
            "Some error has occurred. Check that you confirmed the transaction in MetaMask, and that the contract name is unique"
          );
        }
        setIsSubmitting(false);
      })}
      data-testid="form"
      id="create-xathon"
    >
      <h3 className="font-semibold text-xl self-center">
        Create a new <span className="accent-text">Xathon</span> contract
      </h3>
      <ErrorText className="text-md">{error}</ErrorText>
      <Input
        type="text"
        {...register("address", {
          required: "*Address is required",
          pattern: {
            value: /0x[a-fA-F0-9]{40}/,
            message: "*Must by an ethereum address",
          },
        })}
        errorText={errors.address && errors.address.message}
        labelText="Donation recipient address"
        placeholder="Enter address"
        width="w-2/3"
      />
      <Input
        type="text"
        {...register("name", {
          required: "*Name is required",
          maxLength: {
            value: 50,
            message: "*Must be less than 50 characters in length",
          },
        })}
        errorText={errors.name && errors.name.message}
        labelText="Xathon name"
        placeholder="Enter name"
        width="w-2/3"
      />
      <Textarea
        placeholder="Enter description"
        className="block w-2/3"
        {...register("description")}
      />
      <Input
        type="text"
        {...register("unit", {
          maxLength: {
            value: 20,
            message: "*Must be less than 20 characters in length",
          },
        })}
        labelText="Unit for xathon"
        errorText={errors.unit && errors.unit.message}
        placeholder="miles"
        width="w-30"
      />
      <Input
        type="text"
        {...register("minimumDeposit", { valueAsNumber: true })}
        errorText={errors.minimumDeposit && errors.minimumDeposit.message}
        labelText="Minimum deposit"
        placeholder="0"
        width="w-20"
      />
      <SubmitButton form="create-xathon" className="self-center">
        {isSubmitting ? (
          <span>
            <CgSpinner className="h-5 w-5 mr-1 inline animate-spin" />
            Submitting
          </span>
        ) : (
          "Create"
        )}
      </SubmitButton>
      <p className="text-sm self-center mt-1">
        Your request could take a while to complete
      </p>
    </form>
  );
};

export default CreateXathon;
