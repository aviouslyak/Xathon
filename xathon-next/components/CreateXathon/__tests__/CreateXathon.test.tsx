import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import CreateXathon from "../CreateXathon";
import XathonFactory from "../../../services/contracts/xathonFactory";
import { TransactionReceipt } from "web3-core";
import Router from "next/router";

jest.mock("../../../services/contracts/xathonFactory");
const ADDRESS = "0x0000000000000000000000000000000000000000";

describe("foo", () => {
  let form: HTMLElement;
  let submit: HTMLElement;
  let addressInput: HTMLElement;
  let nameInput: HTMLElement;
  let descriptionInput: HTMLElement;
  let unitInput: HTMLElement;
  let minimumDepositInput: HTMLElement;
  let mockSetAddress: jest.Mock;
  let spiedDeployXathon: jest.SpyInstance<Promise<TransactionReceipt>>;
  let spiedRouterPush: jest.SpyInstance;

  const fillAndSubmit = ({
    address = ADDRESS,
    name = "Foo",
    description = "Foo",
    unit = "miles",
    minimumDeposit = 0,
  }: {
    address?: string | null;
    name?: string | null;
    description?: string | null;
    unit?: string | null;
    minimumDeposit?: number | null;
  }) => {
    address && fireEvent.change(addressInput, { target: { value: address } });
    name && fireEvent.change(nameInput, { target: { value: name } });
    description &&
      fireEvent.change(descriptionInput, { target: { value: description } });
    unit && fireEvent.change(unitInput, { target: { value: unit } });
    minimumDeposit &&
      fireEvent.change(minimumDepositInput, {
        target: { value: minimumDeposit },
      });
    fireEvent.click(submit);
  };
  beforeEach(() => {
    render(<CreateXathon />);
    form = screen.getByTestId("form");
    addressInput = screen.getByLabelText("Donation recipient address");
    nameInput = screen.getByLabelText("Xathon name");
    descriptionInput = screen.getByLabelText("Description");
    unitInput = screen.getByLabelText("Unit for xathon");
    minimumDepositInput = screen.getByLabelText("Minimum deposit");
    submit = screen.getByRole("button");

    spiedDeployXathon = jest.spyOn(XathonFactory.prototype, "deployXathon");
    spiedRouterPush = jest.spyOn(Router, "push");
  });

  test("matches snapshot", () => {
    expect(form).toMatchSnapshot();
  });

  test("address field is required", async () => {
    fillAndSubmit({ address: null });
    await waitFor(() => {
      const errorMessage = screen.getByText("*Address is required");
      expect(errorMessage).toBeInTheDocument();
    });
  });

  test("address field must be of type ethereum address", async () => {
    fillAndSubmit({ address: "not valid" });

    await waitFor(() => {
      const errorMessage = screen.getByText("*Must by an ethereum address");
      expect(errorMessage).toBeInTheDocument();
    });
  });

  test("name field is required", async () => {
    fillAndSubmit({ name: null });

    await waitFor(() => {
      const errorMessage = screen.getByText("*Name is required");
      expect(errorMessage).toBeInTheDocument();
    });
  });

  test("name field can be less than 50 characters", async () => {
    fillAndSubmit({ name: "a".repeat(30) });

    await waitFor(() => {
      const errorMessage = screen.queryByText("*Name is required");
      expect(errorMessage).not.toBeInTheDocument();
      expect(spiedDeployXathon).toHaveBeenCalled();
    });
  });

  test("name field can be 50 characters", async () => {
    fillAndSubmit({ name: "a".repeat(50) });

    await waitFor(() => {
      const errorMessage = screen.queryByText("*Name is required");
      expect(errorMessage).not.toBeInTheDocument();
      expect(spiedDeployXathon).toHaveBeenCalled();
    });
  });

  test("name field can not be more than 50 characters", async () => {
    fillAndSubmit({ name: "a".repeat(51) });

    await waitFor(() => {
      const errorMessage = screen.getByText(
        "*Must be less than 50 characters in length"
      );
      expect(errorMessage).toBeInTheDocument();
    });
  });

  test("description field is not required", async () => {
    fillAndSubmit({ description: null });

    await waitFor(() => {
      expect(XathonFactory.prototype.deployXathon).toHaveBeenCalledTimes(1);
    });
  });

  test("unit field can be less than 20 characters", async () => {
    fillAndSubmit({ unit: "a".repeat(10) });

    await waitFor(() => {
      const errorMessage = screen.queryByText(
        "*Must be less than 20 characters in length"
      );
      expect(errorMessage).not.toBeInTheDocument();
      expect(XathonFactory.prototype.deployXathon).toHaveBeenCalledTimes(1);
    });
  });

  test("unit field can be 20 characters", async () => {
    fillAndSubmit({ unit: "a".repeat(20) });
    await waitFor(() => {
      const errorMessage = screen.queryByText(
        "*Must be less than 20 characters in length"
      );
      expect(errorMessage).not.toBeInTheDocument();
      expect(XathonFactory.prototype.deployXathon).toHaveBeenCalledTimes(1);
    });
  });

  test("unit field can not be more than 20 characters", async () => {
    fillAndSubmit({ unit: "a".repeat(21) });

    await waitFor(() => {
      const errorMessage = screen.getByText(
        "*Must be less than 20 characters in length"
      );
      expect(errorMessage).toBeInTheDocument();
    });
  });

  test("minimumDeposit field is not required", async () => {
    fillAndSubmit({ minimumDeposit: null });

    await waitFor(() => {
      expect(XathonFactory.prototype.deployXathon).toHaveBeenCalledTimes(1);
    });
  });

  test("successfully calls deployXathon with correct values", async () => {
    fillAndSubmit({
      name: "N",
      description: "D",
      unit: "M",
      minimumDeposit: 0,
    });
    await waitFor(() => {
      expect(XathonFactory.prototype.deployXathon).toHaveBeenCalledTimes(1);
      expect(XathonFactory.prototype.deployXathon).toHaveBeenCalledWith({
        address: ADDRESS,
        name: "N",
        description: "D",
        unit: "M",
        minimumDeposit: 0,
      });
    });
  });

  test("minimumDeposit has correct default value", () => {
    const minimumDepositRefound = screen.getByDisplayValue(0);
    expect(minimumDepositRefound).toBeInTheDocument();
  });

  test("description, name, unit, and address have default values", () => {
    const defaults = screen.getAllByDisplayValue("");
    expect(defaults.length).toBe(4);
  });

  test("redirects after submitting", async () => {
    fillAndSubmit({});

    await waitFor(() => {
      expect(spiedRouterPush).toHaveBeenCalled();
      expect(spiedRouterPush).toHaveBeenCalledWith(`/${ADDRESS}`);
    });
  });

  test("sets loading text when loading", async () => {
    fillAndSubmit({});
    await waitFor(() => {
      const buttonRefound = screen.getByText(/submitting/i);
      expect(buttonRefound).toBeInTheDocument();
    });
  });

  test("error is shown when deployXathon fails", async () => {
    spiedDeployXathon.mockRejectedValueOnce("foo");

    fillAndSubmit({});

    await waitFor(() => {
      const errorText = screen.getByText(
        "Some error has occurred. Check that you confirmed the transaction in MetaMask, and that the contract name is unique"
      );
      expect(errorText).toBeInTheDocument();
    });
  });

  test("error is not shown by default", () => {
    const errorText = screen.queryByText(
      "Some error has occurred. Check that you confirmed the transaction in MetaMask, and that the contract name is unique"
    );
    expect(errorText).not.toBeInTheDocument();
  });
});
