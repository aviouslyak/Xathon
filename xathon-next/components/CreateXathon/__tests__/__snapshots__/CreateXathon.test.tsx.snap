// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`foo matches snapshot 1`] = `
<form
  class="w-1/2 p-4 flex flex-col justify-items-start items-start gap-1"
  data-testid="form"
  id="create-xathon"
>
  <h3
    class="font-semibold text-xl self-center"
  >
    Create a new 
    <span
      class="accent-text"
    >
      Xathon
    </span>
     contract
  </h3>
  <p
    class="text-black text-md"
  />
  <label
    class="font-semibold"
    for="address"
  >
    Donation recipient address
  </label>
  <p
    class="text-black text-md"
  />
  <div
    class="w-2/3 h-10 relative"
    data-testid="input-wrapper"
  >
    <input
      class="rounded-lg bg-pink-200 focus:outline-none focus:border-2 border-pink-500 shadow-md shadow-pink-800  h-full w-full pl-2 pr-2  "
      id="address"
      name="address"
      placeholder="Enter address"
      type="text"
    />
  </div>
  <label
    class="font-semibold"
    for="name"
  >
    Xathon name
  </label>
  <p
    class="text-black text-md"
  />
  <div
    class="w-2/3 h-10 relative"
    data-testid="input-wrapper"
  >
    <input
      class="rounded-lg bg-pink-200 focus:outline-none focus:border-2 border-pink-500 shadow-md shadow-pink-800  h-full w-full pl-2 pr-2  "
      id="name"
      name="name"
      placeholder="Enter name"
      type="text"
    />
  </div>
  <label
    class="font-semibold"
    for="description"
  >
    Description
  </label>
  <p
    class="text-black text-xs"
  />
  <textarea
    class="
          resize-none rounded-lg block px-2 bg-pink-200 focus:outline-none focus:border-2 border-pink-500 shadow-md shadow-pink-800 block w-2/3 h-20"
    id="description"
    name="description"
    placeholder="Enter description"
  />
  <label
    class="font-semibold"
    for="unit"
  >
    Unit for xathon
  </label>
  <p
    class="text-black text-md"
  />
  <div
    class="w-30 h-10 relative"
    data-testid="input-wrapper"
  >
    <input
      class="rounded-lg bg-pink-200 focus:outline-none focus:border-2 border-pink-500 shadow-md shadow-pink-800  h-full w-full pl-2 pr-2  "
      id="unit"
      name="unit"
      placeholder="miles"
      type="text"
    />
  </div>
  <label
    class="font-semibold"
    for="minimumDeposit"
  >
    Minimum deposit
  </label>
  <p
    class="text-black text-md"
  />
  <div
    class="w-20 h-10 relative"
    data-testid="input-wrapper"
  >
    <input
      class="rounded-lg bg-pink-200 focus:outline-none focus:border-2 border-pink-500 shadow-md shadow-pink-800  h-full w-full pl-2 pr-2  "
      id="minimumDeposit"
      name="minimumDeposit"
      placeholder="0"
      type="text"
    />
  </div>
  <button
    class="self-center group w-40 h-10 text-xl font-semibold flex items-center justify-center rounded-lg transition-shadow shadow-md hover:shadow-lg shadow-pink-900 hover:shadow-pink-900 bg-orange-400"
    form="create-xathon"
    type="submit"
  >
    <span>
      Create
      <svg
        class="transition-transform transform group-hover:scale-125 ml-2 inline-block"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 1024 1024"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M696 480H544V328c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v152H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8h152v152c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V544h152c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8z"
        />
        <path
          d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
        />
      </svg>
    </span>
  </button>
  <p
    class="text-sm self-center mt-1"
  >
    Your request could take a while to complete
  </p>
</form>
`;
